### PR TITLE
Allow overriding the AsyncMethodBuilder on methods

### PR DESCRIPTION
**Pull request by jcouv**
Spec: https://github.com/dotnet/csharplang/blob/main/proposals/async-method-builders.md
Test plan: https://github.com/dotnet/roslyn/issues/51999

Current design choices:
- we don't need the return type of async methods with AsyncMethodBuilder to be proper task-like types
- we do not care about the accessibility of the builder override
- but all members of the builder overrides (factory type and final builder) must have `public` accessibility
- there is no builder factory vs. builder indirection after all
- the AsyncMethodBuilder is only allowed on lambdas when their return type is explicit

### CODE REVIEW

File: src/Compilers/CSharp/Portable/Symbols/MethodSymbolExtensions.cs
Comment (cston) [inline src/Compilers/CSharp/Portable/Symbols/MethodSymbolExtensions.cs:215]: >HasMethodLevelBuilder(this MethodSymbol method [](http://example.com/codeflow?start=29&length=46)

Consider changing this to an extension method on `Symbol` and call this method from `TypeSymbolExtensions.IsCustomTaskType()` as well.

File: src/Compilers/CSharp/Portable/Symbols/MethodSymbolExtensions.cs
Comment (cston) [inline src/Compilers/CSharp/Portable/Symbols/MethodSymbolExtensions.cs:195]: >IsAsyncReturningTaskViaOverride [](http://example.com/codeflow?start=27&length=31)

Are `IsAsyncReturningGenericTask()` and `IsAsyncReturningTaskViaOverride()` ever called separately? If not, consider combining. Same question for generic task pair.

File: src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs
Comment (cston) [inline src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs:2658]: Are these two methods ever called separately? If not, consider combining the methods. Same question for pair of methods below.

**Comment by cston** [comment]
>         public void BuilderFactoryOnMethod_CreateHasParamaeter()

Typo.

---
Refers to: src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenAsyncMethodBuilderOverrideTests.cs:991 in e1f1d41. [](commit_id = e1f1d413b3de8bed6daaed5e314bf9402404259e, deletion_comment = False)

File: src/Compilers/CSharp/Portable/Symbols/MethodSymbolExtensions.cs
Comment (jcouv) [reply src/Compilers/CSharp/Portable/Symbols/MethodSymbolExtensions.cs:215]: I'm looking at one usage of `IsCustomTaskType` in `BoundLambda.CalculateReturnType` which I need to think more about.

File: src/Compilers/CSharp/Portable/BoundTree/UnboundLambda.cs
Comment (jcouv) [inline src/Compilers/CSharp/Portable/BoundTree/UnboundLambda.cs:680]: :memo: Changes to this method are covered by `BuilderFactoryOnMethod_OnLambda_NotTaskLikeTypes`. With this change the test doesn't report `ERR_AnonymousReturnExpected` nor `ERR_CantConvAsyncAnonFuncReturns` below.

File: src/Compilers/CSharp/Portable/Lowering/AsyncRewriter/AsyncMethodBuilderMemberCollection.cs
Comment (cston) [inline src/Compilers/CSharp/Portable/Lowering/AsyncRewriter/AsyncMethodBuilderMemberCollection.cs:388]: Could we use the existing `TryCreate()` method instead?

Comment (jcouv) [reply src/Compilers/CSharp/Portable/Lowering/AsyncRewriter/AsyncMethodBuilderMemberCollection.cs:388]: It doesn't make much difference. I did initially ;-) I'll change it back.

File: src/Compilers/CSharp/Portable/Symbols/MethodSymbolExtensions.cs
Comment (cston) [reply src/Compilers/CSharp/Portable/Symbols/MethodSymbolExtensions.cs:199]: I didn't understand the comment about `CalculateReturnType`. Could we share this implementation with `TypeSymbolExtensions.IsCustomTaskType()`?

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceMethodSymbolWithAttributes.cs
Comment (cston) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceMethodSymbolWithAttributes.cs:1061]: >hasErrors [](http://example.com/codeflow?start=20&length=9)

It looks like this may reset `hasErrors` that was set above.

File: src/Compilers/CSharp/Portable/Symbols/MethodSymbolExtensions.cs
Comment (jcouv) [reply src/Compilers/CSharp/Portable/Symbols/MethodSymbolExtensions.cs]: The comment about `CalculateReturnType` is no longer relevant now that we're checking the builder from the return type.
Factored logic with `IsCustomTaskType`.

**Comment by jcouv** [comment]
@dotnet/roslyn-compiler for a second review. Let me know if you need a walkthrough. Thanks

File: src/Compilers/CSharp/Portable/Symbols/Source/LambdaSymbol.cs
Comment (cston) [inline src/Compilers/CSharp/Portable/Symbols/Source/LambdaSymbol.cs:284]: >HasAsyncMethodBuilder [](http://example.com/codeflow?start=21&length=21)

It may be redundant, but consider checking `MessageID.IDS_AsyncMethodBuilderOverride.CheckFeatureAvailability()` here.

File: src/Compilers/CSharp/Test/Semantic/Semantics/BindingAsyncTasklikeMoreTests.cs
Comment (cston) [inline src/Compilers/CSharp/Test/Semantic/Semantics/BindingAsyncTasklikeMoreTests.cs:166]: >AsyncMethod_NullBuilder [](http://example.com/codeflow?start=20&length=23)

It looks like we're already testing this case in `AsyncTasklikeBadAttributeArgument4()`.

**Comment by cston** [comment]
>             static string asyncBuilderFactoryCode(string builderTypeName, string tasklikeTypeName, string? genericTypeParameter = null, bool isStruct = false)

Is this parameter ever set?

---
Refers to: src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenAsyncMethodBuilderOverrideTests.cs:722 in ee207cb. [](commit_id = ee207cb9a47d181c2552a94f272b960f3b728b25, deletion_comment = False)

File: src/Compilers/CSharp/Portable/Symbols/Source/LambdaSymbol.cs
Comment (jcouv) [reply src/Compilers/CSharp/Portable/Symbols/Source/LambdaSymbol.cs:284]: After I merge with explicit return types that won't be needed, since you've added a call to `AsyncMethodChecks` here.

**Comment by jcouv** [comment]
>     [AsyncMethodBuilder(typeof(MyTaskMethodBuilder))]

I didn't test any override scenarios since there was no code change related to that.

---
In reply to: [865423661](https://github.com/dotnet/roslyn/pull/54033#issuecomment-865423661) [](http://example.com/codeflow?ancestors=865423661,865423661)

---
Refers to: src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenAsyncMethodBuilderOverrideTests.cs:95 in ee207cb. [](commit_id = ee207cb9a47d181c2552a94f272b960f3b728b25, deletion_comment = False)

**Comment by jcouv** [comment]
Merged with explicit return types. We'll have to merge https://github.com/dotnet/roslyn/pull/54282 before continuing the review.

**Comment by cston** [comment]
>     [AsyncMethodBuilder(typeof(MyTaskMethodBuilder))]

My mistake - the attribute does not affect callers.

---
In reply to: [865452212](https://github.com/dotnet/roslyn/pull/54033#issuecomment-865452212) [](http://example.com/codeflow?ancestors=865452212,865423661)

---
Refers to: src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenAsyncMethodBuilderOverrideTests.cs:95 in ee207cb. [](commit_id = ee207cb9a47d181c2552a94f272b960f3b728b25, deletion_comment = False)

**Comment by jcouv** [comment]
@dotnet/roslyn-compiler for review. Thanks

**Comment by cston** [comment]
>                 // (7,71): error CS8934: The AsyncMethodBuilder attribute is disallowed on anonymous methods without an explicit return type.

CS8935 here and in other locations.

---
Refers to: src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenAsyncMethodBuilderOverrideTests.cs:889 in 3268386. [](commit_id = 3268386cadaa342ffccca9b936a493ea53c3d8e6, deletion_comment = False)

**Review by cston — APPROVED** [review]

File: src/Compilers/CSharp/Portable/Lowering/AsyncRewriter/AsyncMethodBuilderMemberCollection.cs
Comment (333fred) [inline src/Compilers/CSharp/Portable/Lowering/AsyncRewriter/AsyncMethodBuilderMemberCollection.cs:478]: Is this actually just a bug fix?

File: src/Compilers/CSharp/Portable/Symbols/SymbolExtensions.cs
Comment (333fred) [inline src/Compilers/CSharp/Portable/Symbols/SymbolExtensions.cs:820]: >the [](http://example.com/codeflow?start=105&length=3)

Nit: remove `the`

Comment (333fred) [reply src/Compilers/CSharp/Portable/Symbols/SymbolExtensions.cs:820]: or add `type`

File: src/Compilers/CSharp/Portable/Lowering/AsyncRewriter/AsyncMethodBuilderMemberCollection.cs
Comment (jcouv) [reply src/Compilers/CSharp/Portable/Lowering/AsyncRewriter/AsyncMethodBuilderMemberCollection.cs:478]: We need this change for proper behavior of method-level builders (see test `BuilderOnMethod_CreateHasRefReturn`).
But yes it was a misbehavior of task-like types (see test `AsyncMethod_CreateHasRefReturn`).

File: src/Compilers/CSharp/Portable/Symbols/SymbolExtensions.cs
Comment (333fred) [inline src/Compilers/CSharp/Portable/Symbols/SymbolExtensions.cs:834]: >ValueInternal [](http://example.com/codeflow?start=73&length=13)

If we know this is a type, why not have the builderArgument be a TypeSymbol?

**Comment by 333fred** [comment]
>         internal static bool IsCustomTaskType(this NamedTypeSymbol type, [NotNullWhen(true)] out object? builderArgument)

Same question here, I guess.

---
Refers to: src/Compilers/CSharp/Portable/Symbols/TypeSymbolExtensions.cs:1643 in 3268386. [](commit_id = 3268386cadaa342ffccca9b936a493ea53c3d8e6, deletion_comment = False)

File: src/Compilers/CSharp/Portable/Symbols/SymbolExtensions.cs
Comment (333fred) [inline src/Compilers/CSharp/Portable/Symbols/SymbolExtensions.cs:823]: >HasAsyncMethodBuilder [](http://example.com/codeflow?start=29&length=21)

Nit: feels like a better name for this is `HasAsyncMethodBuilderAttribute`, since it specifically won't return the builder type if the method has a Task-like type with no attribute.

**Comment by 333fred** [comment]
>                 // (75,19): error CS1983: The return type of an async method must be void, Task, Task<T>, a task-like type, IAsyncEnumerable<T>, or IAsyncEnumerator<T>

These errors have nothing to do with accessibility? Also, several of these have no issues?

---
Refers to: src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenAsyncMethodBuilderOverrideTests.cs:662 in 3268386. [](commit_id = 3268386cadaa342ffccca9b936a493ea53c3d8e6, deletion_comment = False)

**Comment by 333fred** [comment]
>     [AsyncMethodBuilder(typeof(MyTaskMethodBuilder))] static async MyTask () => {{ System.Console.Write(""Lambda1 ""); await Task.Delay(0); }} // 1

I would have expected this to error?

---
Refers to: src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenAsyncMethodBuilderOverrideTests.cs:918 in 3268386. [](commit_id = 3268386cadaa342ffccca9b936a493ea53c3d8e6, deletion_comment = False)

**Comment by 333fred** [comment]
Done review pass (commit 25)

**Comment by jcouv** [comment]
>     [AsyncMethodBuilder(typeof(MyTaskMethodBuilder))] static async MyTask () => {{ System.Console.Write(""Lambda1 ""); await Task.Delay(0); }} // 1

Resolved offline

---
In reply to: [867139447](https://github.com/dotnet/roslyn/pull/54033#issuecomment-867139447) [](http://example.com/codeflow?ancestors=867139447,867139447)

---
Refers to: src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenAsyncMethodBuilderOverrideTests.cs:918 in 3268386. [](commit_id = 3268386cadaa342ffccca9b936a493ea53c3d8e6, deletion_comment = False)

**Comment by jcouv** [comment]
>         public void BuilderOnMethod_WrongAccessibility()

Will fix this test to have builder on method and adjust name.

---
Refers to: src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenAsyncMethodBuilderOverrideTests.cs:632 in 9440d0f. [](commit_id = 9440d0ff08c79370e516386c7942047759e932be, deletion_comment = False)

File: src/Compilers/CSharp/Portable/Symbols/SymbolExtensions.cs
Comment (jcouv) [reply src/Compilers/CSharp/Portable/Symbols/SymbolExtensions.cs:834]: Kept in line with existing `IsCustomTaskType` method.

**Review by 333fred — APPROVED** [review]

**Pull request merged**
